File: backend/sdk/handlers/location_tools.py
"""
Location management tools for TRPG gameplay.

Contains tools for location navigation and management:
- persist_location_design: Create location (used by sub-agents via Task tool)
- travel: Move player to an existing location
- list_locations: List all available locations

"""

import asyncio
import logging
from typing import Any, Optional

import crud
import schemas
from claude_agent_sdk import tool
from domain.value_objects.enums import MessageRole
from infrastructure.logging.perf_logger import track_perf
from services.location_service import LocationService
from services.persistence_manager import PersistenceManager
from services.player_service import PlayerService
from services.world_service import WorldService

from sdk.handlers.common import build_action_context
from sdk.handlers.context import ToolContext
from sdk.loaders import get_tool_description, is_tool_enabled
from sdk.tools.gameplay import (
    ListLocationsInput,
    TravelInput,
)
from sdk.tools.subagent import PersistLocationDesignInput

logger = logging.getLogger("GameplayTools.Location")


def create_location_tools(ctx: ToolContext) -> list:
    """
    Create location management tools.

    Args:
        ctx: Unified tool context containing agent info and dependencies

    Returns:
        List of location tool functions
    """
    tools = []

    # Get required dependencies from context
    db = ctx.require_db()
    world_id = ctx.require_world_id()
    world_name = ctx.require_world_name()

    # ==========================================================================
    # travel tool - Move player to existing location only
    # ==========================================================================
    if is_tool_enabled("travel", default=True):
        travel_description = get_tool_description(
            "travel",
            agent_name="Action Manager",
            group_name=ctx.group_name,
            default="Move the player to an existing location. Use Task tool with location_designer first if the destination doesn't exist. Use bring_characters to bring characters along.",
        )

        @tool(
            "travel",
            travel_description,
            TravelInput.model_json_schema(),
        )
        @track_perf(
            "tool_travel",
            room_id=lambda: ctx.room_id,
            agent_name=lambda: ctx.agent_name,
            include_result=True,
        )
        async def travel_tool(args: dict[str, Any]):
            """Move the player to an existing location with narration, action suggestions, and chat summary."""
            # Validate input with Pydantic
            validated = TravelInput(**args)
            destination = validated.destination
            bring_characters = validated.bring_characters
            narration = validated.narration
            action_1 = validated.action_1
            action_2 = validated.action_2
            chat_summary = validated.chat_summary
            user_action = validated.user_action

            if not destination:
                return {
                    "content": [{"type": "text", "text": "No destination specified."}],
                    "is_error": True,
                }

            logger.info(f"travel invoked: '{destination}' (bringing={bring_characters})")

            # Get current room_id for sub-agent status display
            # Import here to avoid circular import
            from orchestration import get_trpg_orchestrator

            trpg_orchestrator = get_trpg_orchestrator()
            current_room_id: Optional[int] = None

            try:
                # Get current location for summarization
                context = build_action_context(world_name, f"travel to {destination}")
                from_location_name = context.current_location

                # Get current location's room_id for status display
                if from_location_name and from_location_name != "unknown":
                    from_loc = await crud.get_location_by_name(db, world_id, from_location_name)
                    if from_loc and from_loc.room_id:
                        current_room_id = from_loc.room_id  # type: ignore[assignment]  # SQLAlchemy Column coerces to int at runtime
                        # Set sub-agent status to show "Traveling to..." in frontend
                        trpg_orchestrator.set_sub_agent_active(
                            current_room_id,
                            "Travel",
                            f"Traveling to {destination}...",
                        )

                # Save chat summary to world history (using provided summary, no sub-agent needed)
                if from_location_name and from_location_name != "unknown":
                    try:
                        # Get location display name for history entry
                        from_location_config = LocationService.load_location(world_name, from_location_name)
                        from_display_name = (
                            from_location_config.display_name if from_location_config else from_location_name
                        )

                        # Get current turn count from DB (DB has the incremented value, FS may not be synced)
                        db_player_state = await crud.get_player_state(db, world_id)
                        turn = db_player_state.turn_count if db_player_state else 0

                        # Save the provided chat summary to history
                        WorldService.add_history_entry(world_name, turn, from_display_name, chat_summary)
                        logger.info(f"Saved chat summary to history: {from_display_name} (turn {turn})")
                    except Exception as e:
                        logger.warning(f"Failed to save chat summary to history: {e}")

                    # Trigger memory round for NPCs at the departing location
                    try:
                        if from_loc and from_loc.id:
                            npc_count = await trpg_orchestrator.trigger_npc_memory_round(from_loc.id)
                            if npc_count > 0:
                                logger.info(f"Memory round complete: {npc_count} NPCs processed")
                    except Exception as e:
                        logger.warning(f"Failed to trigger memory round: {e}")

                # Get all existing locations
                db_locations = await crud.get_locations(db, world_id)

                # Try to find matching location (case-insensitive)
                destination_lower = destination.lower()
                matching_location = None

                for loc in db_locations:
                    # Match by name or display_name (case-insensitive)
                    if (
                        loc.name.lower() == destination_lower
                        or (loc.display_name and loc.display_name.lower() == destination_lower)
                        or destination_lower in loc.name.lower()
                        or (loc.display_name and destination_lower in loc.display_name.lower())
                    ):
                        matching_location = loc
                        break

                # Track destination location ID for character movement
                destination_location_id = None
                display_name = ""
                pos_x, pos_y = 0, 0

                # Get from_location_id for character movement
                from_location = next(
                    (loc for loc in db_locations if loc.name == from_location_name),
                    None,
                )
                from_location_id = from_location.id if from_location else None

                # If no match found, return error - use add_location to create new locations
                if matching_location is None:
                    # Get all location display names for helpful error message
                    location_names = [loc.display_name or loc.name for loc in db_locations]
                    available = ", ".join(location_names) if location_names else "none"
                    logger.warning(f"Location '{destination}' not found. Available: {available}")
                    return {
                        "content": [
                            {
                                "type": "text",
                                "text": f"Location '{destination}' does not exist. Use Task tool with location_designer to create it first.\n\nAvailable locations: {available}",
                            }
                        ],
                        "is_error": True,
                    }
                else:
                    # Move to existing location - create fresh room for new visit
                    destination_location_id = matching_location.id

                    # Create new room for this visit (clean context)
                    new_room = await crud.create_new_room_for_location(db, matching_location)
                    logger.info(f"Created new room {new_room.id} for location {matching_location.name}")

                    # Update room mapping in _state.json
                    room_key = LocationService.location_to_room_key(matching_location.name)
                    LocationService.set_room_mapping(
                        world_name=world_name,
                        room_key=room_key,
                        db_room_id=new_room.id,
                        agents=[],  # Gameplay agents added via DB
                    )

                    # Pre-connect characters at destination (max 5) for faster NPC reactions
                    if ctx.agent_manager and destination_location_id:

                        async def pre_connect_destination_chars():
                            try:
                                dest_chars = await crud.get_characters_at_location(
                                    db, destination_location_id, exclude_system_agents=True
                                )
                                # Limit to 5 characters to avoid resource exhaustion
                                for char in dest_chars[:5]:
                                    await ctx.agent_manager.pre_connect(
                                        db=db,
                                        room_id=new_room.id,
                                        agent_id=char.id,
                                        agent_name=char.name,
                                        world_name=world_name,
                                        world_id=world_id,
                                        config_file=char.config_file,
                                        group_name=char.group,
                                    )
                            except Exception as e:
                                logger.debug(f"Pre-connect chars failed (non-critical): {e}")

                        # Fire-and-forget background task
                        asyncio.create_task(pre_connect_destination_chars())

                    await crud.set_current_location(db, world_id, matching_location.id)

                    # Update filesystem player state
                    fs_state = PlayerService.load_player_state(world_name)
                    if fs_state:
                        fs_state.current_location = matching_location.name
                        PlayerService.save_player_state(world_name, fs_state)

                    display_name = matching_location.display_name or matching_location.name
                    pos_x, pos_y = matching_location.position_x, matching_location.position_y

                    logger.info(f"Traveled to existing location: {display_name}")

                # ============================================================
                # Move accompanying characters
                # ============================================================
                moved_characters = []
                for char_name in bring_characters:
                    try:
                        agent = await crud.get_agent_by_name(db, char_name)
                        if agent and destination_location_id:
                            await crud.move_character_to_location(
                                db, agent.id, from_location_id, destination_location_id
                            )
                            moved_characters.append(char_name)
                            logger.info(f"Moved character {char_name} to {display_name}")
                        else:
                            logger.warning(f"Character '{char_name}' not found, skipping")
                    except Exception as e:
                        logger.warning(f"Failed to move character '{char_name}': {e}")

                # ============================================================
                # Create narrative message in the NEW room
                # ============================================================
                try:
                    agent_id = ctx.require_agent_id()
                    # Use the new room for the narration message
                    new_room_id = new_room.id if new_room else ctx.require_room_id()

                    message = schemas.MessageCreate(
                        content=narration,
                        role=MessageRole.ASSISTANT,
                        agent_id=agent_id,
                    )
                    await crud.create_message(db, new_room_id, message, update_room_activity=True)
                    logger.info(f"Travel narration created | room={new_room_id} | agent={agent_id}")
                except Exception as e:
                    logger.error(f"Failed to create travel narration: {e}")

                # ============================================================
                # Save action suggestions to _state.json
                # ============================================================
                try:
                    LocationService.save_suggestions(world_name, [action_1, action_2])
                    logger.info(f"Travel suggestions saved: [{action_1}] / [{action_2}]")
                except Exception as e:
                    logger.error(f"Failed to save travel suggestions: {e}")

                # ============================================================
                # Save arrival context for continuity (one-time use)
                # ============================================================
                try:
                    LocationService.save_arrival_context(
                        world_name=world_name,
                        previous_narration=narration,
                        triggering_action=user_action,
                        from_location=from_location_name or "unknown",
                    )
                    logger.info("Arrival context saved for continuity")
                except Exception as e:
                    logger.error(f"Failed to save arrival context: {e}")

                # Build response
                response_text = f"""**Traveled to:** {display_name}
- Position: ({pos_x}, {pos_y})"""

                if moved_characters:
                    response_text += f"\n- Companions: {', '.join(moved_characters)}"

                response_text += "\n\n Narrative message created and displayed to player."
                response_text += f"\n Suggested actions: [{action_1}] / [{action_2}]"
                response_text += "\n Chat summary saved to world history."
                response_text += "\n Arrival context saved for next action continuity."

                return {"content": [{"type": "text", "text": response_text}]}

            except Exception as e:
                logger.error(f"travel error: {e}", exc_info=True)
                return {
                    "content": [{"type": "text", "text": f"Error during travel: {e}"}],
                    "is_error": True,
                }
            finally:
                # Clear sub-agent status when travel completes (success or error)
                if current_room_id:
                    trpg_orchestrator.set_sub_agent_inactive(current_room_id)

        tools.append(travel_tool)

    # ==========================================================================
    # list_locations tool - List all available locations in the world
    # ==========================================================================
    if is_tool_enabled("list_locations", default=True):
        list_locations_description = get_tool_description(
            "list_locations",
            agent_name="Action Manager",
            group_name=ctx.group_name,
            default="List all available locations in the current world.",
        )

        @tool("list_locations", list_locations_description, ListLocationsInput.model_json_schema())
        @track_perf(
            "tool_list_locations",
            room_id=lambda: ctx.room_id,
            agent_name=lambda: ctx.agent_name,
            include_result=True,
        )
        async def list_locations_tool(_args: dict[str, Any]):
            """List all available locations in the world."""
            logger.info("list_locations invoked")

            try:
                # Load locations from filesystem (source of truth)
                fs_locations = LocationService.load_all_locations(world_name)

                if not fs_locations:
                    return {
                        "content": [{"type": "text", "text": "No locations found in this world."}],
                    }

                # Get current location for highlighting
                state = PlayerService.load_player_state(world_name)
                current_location = state.current_location if state else None

                # Build location list from filesystem locations
                location_entries = []
                for loc_name, loc_config in fs_locations.items():
                    display_name = loc_config.display_name or loc_name
                    entry = f"- **{display_name}**"

                    if loc_name == current_location:
                        entry += " (current)"

                    # Add position info
                    pos = loc_config.position if isinstance(loc_config.position, tuple) else (0, 0)
                    entry += f" at ({pos[0]}, {pos[1]})"

                    # Add description if available
                    if loc_config.description:
                        # Truncate description to first 100 chars
                        desc_preview = loc_config.description[:100]
                        if len(loc_config.description) > 100:
                            desc_preview += "..."
                        entry += f"\n  {desc_preview}"

                    location_entries.append(entry)

                response_text = f"**Locations in {world_name}:**\n\n" + "\n\n".join(location_entries)

                return {"content": [{"type": "text", "text": response_text}]}

            except Exception as e:
                logger.error(f"list_locations error: {e}", exc_info=True)
                return {
                    "content": [{"type": "text", "text": f"Error listing locations: {e}"}],
                    "is_error": True,
                }

        tools.append(list_locations_tool)

    # ==========================================================================
    # persist_location_design tool - For sub-agents to persist location designs
    # ==========================================================================
    if is_tool_enabled("persist_location_design", default=True):
        persist_location_description = get_tool_description(
            "persist_location_design",
            agent_name="Location Designer",
            group_name=ctx.group_name,
            default="Persist a location design to the game world. Creates the location in filesystem and database.",
        )

        @tool(
            "persist_location_design",
            persist_location_description,
            PersistLocationDesignInput.model_json_schema(),
        )
        async def persist_location_design_tool(args: dict[str, Any]):
            """Persist a designed location to filesystem and database.

            Used by Location Designer sub-agent after designing a location.
            Creates the location files and connects to adjacent locations.
            Returns an error if the location already exists.
            """
            validated = PersistLocationDesignInput(**args)

            logger.info(f"persist_location_design: {validated.display_name}")

            try:
                # Check if location already exists
                db_locations = await crud.get_locations(db, world_id)
                location_name_lower = validated.name.lower()

                existing = None
                for loc in db_locations:
                    if loc.name.lower() == location_name_lower or (
                        loc.display_name and loc.display_name.lower() == location_name_lower
                    ):
                        existing = loc
                        break

                if existing:
                    return {
                        "content": [
                            {
                                "type": "text",
                                "text": f"Location '{validated.name}' already exists. Cannot overwrite.",
                            }
                        ],
                        "is_error": True,
                    }

                # Build adjacent hints from adjacent_to (already a list or None)
                adjacent_hints = validated.adjacent_to or []

                # Use PersistenceManager for coordinated FS + DB creation
                pm = PersistenceManager(db, world_id, world_name)
                new_location_id = await pm.create_location(
                    name=validated.name,
                    display_name=validated.display_name,
                    description=validated.description,
                    position=(validated.position_x, validated.position_y),
                    adjacent_hints=adjacent_hints,
                    is_starting=validated.is_starting,
                )

                # Connect to adjacent locations in DB
                if adjacent_hints:
                    for adj_name in adjacent_hints:
                        adj_loc = next((loc for loc in db_locations if loc.name == adj_name), None)
                        if adj_loc:
                            await crud.add_adjacent_location(db, new_location_id, adj_loc.id)
                            await crud.add_adjacent_location(db, adj_loc.id, new_location_id)

                # Note: is_starting is informational only during onboarding.
                # The actual starting_location is set by OM's complete tool.
                if validated.is_starting:
                    logger.info(f"Location '{validated.name}' marked as starting location candidate")

                logger.info(f"Created location: {validated.display_name} (id={new_location_id})")

                response_text = f"""**Location Created:**
- Name: {validated.display_name}
- Position: ({validated.position_x}, {validated.position_y})
- Is Starting: {validated.is_starting}
- Description: {validated.description[:200]}..."""

                return {"content": [{"type": "text", "text": response_text}]}

            except Exception as e:
                logger.error(f"persist_location_design error: {e}", exc_info=True)
                return {
                    "content": [{"type": "text", "text": f"Error creating location: {e}"}],
                    "is_error": True,
                }

        tools.append(persist_location_design_tool)

    return tools

File: backend-ts/src/sdk/handlers/location-tools.ts
import { getAgentByName } from '../../crud/agents'
import {
  addAdjacentLocation,
  createNewRoomForLocation,
  getLocationByName,
  getLocations,
  moveCharacterToLocation,
} from '../../crud/locations'
import { createMessage } from '../../crud/messages'
import { getPlayerState, setCurrentLocation } from '../../crud/player-state'
import type { LocationStorage } from '../../services/location-storage'
import type { PlayerService } from '../../services/player-service'
import { RoomMappingService } from '../../services/room-mapping'
import type { WorldService } from '../../services/world-service'
import { travelTool } from '../tools/location'
import { resolveTool } from '../tools/registry'
import { persistLocationDesignTool } from '../tools/subagent'
import { getLogger } from '../../infrastructure/logging/logger'
import type { LocationPersistenceFactory, TurnStatusPort } from './ports'
import { tool, requireAgentId, requireRoomId, requireWorldId, requireWorldName, toolError, toolSuccess, type SdkTool, type ToolContext } from './context'

/**
 * Moving the player, and creating somewhere to move them to.
 * Port of `sdk/handlers/location_tools.py`.
 *
 * `list_locations` is not here — Phase 0 put it in `world-tools.ts`.
 *
 * **`travel` is the widest side effect in the game.** One call ends the scene,
 * writes its summary to world history, gives the departing NPCs a memory round,
 * opens a *new room* at the destination, moves the player and their companions,
 * writes the arrival narration into that new room, saves the next turn's
 * suggested actions and parks a continuity payload for the following turn. It
 * is one tool rather than six because every one of those has to happen, in that
 * order, or the player arrives somewhere with no narration and no way to act.
 */

const logger = getLogger('GameplayTools.Location')

export interface LocationToolDeps {
  players: PlayerService
  rooms: RoomMappingService
  locations: LocationStorage
  worlds: WorldService
  status?: TurnStatusPort
  /** Required by `persist_location_design`; absent means the tool is not offered. */
  persistence?: LocationPersistenceFactory
}

export function createLocationTools(ctx: ToolContext, deps: LocationToolDeps): SdkTool[] {
  const worldName = requireWorldName(ctx)
  const worldId = requireWorldId(ctx)

  const def = resolveTool(travelTool.name, ctx.groupName)
  if (!def) return []

  const travel = tool(
    travelTool.name,
    def.description,
    travelTool.inputSchema,
    async (args) => {
      const {
        destination,
        bring_characters: bringCharacters,
        narration,
        action_1: action1,
        action_2: action2,
        chat_summary: chatSummary,
        user_action: userAction,
      } = args

      const db = ctx.getDb()
      // Held across the whole call so the `finally` can clear the indicator on
      // the *departure* room, which is the room the player is still watching.
      let departureRoomId: number | null = null

      try {
        const fromLocationName = deps.players.loadPlayerState(worldName)?.currentLocation || 'unknown'
        const fromLocation =
          fromLocationName === 'unknown' ? null : getLocationByName(db, worldId, fromLocationName)

        if (fromLocation?.roomId) {
          departureRoomId = fromLocation.roomId
          deps.status?.setSubAgentActive?.(
            departureRoomId,
            'Travel',
            `Traveling to ${destination}...`,
          )
        }

        if (fromLocationName !== 'unknown') {
          try {
            const config = deps.locations.loadLocation(worldName, fromLocationName)
            const fromDisplay = config?.displayName ?? fromLocationName
            // The database's turn count, not the filesystem's: `incrementTurn`
            // writes the row and `player.yaml` may not have caught up yet, so
            // reading the file here numbers the history entry one turn behind.
            const turn = getPlayerState(db, worldId)?.turnCount ?? 0
            deps.worlds.addHistoryEntry(worldName, turn, fromDisplay, chatSummary)
          } catch (error) {
            logger.warning(`Failed to save chat summary to history: ${String(error)}`)
          }

          try {
            if (fromLocation) {
              const count = (await deps.status?.triggerNpcMemoryRound?.(fromLocation.id)) ?? 0
              if (count > 0) logger.info(`Memory round complete: ${count} NPCs processed`)
            }
          } catch (error) {
            logger.warning(`Failed to trigger memory round: ${String(error)}`)
          }
        }

        const dbLocations = getLocations(db, worldId)
        const destinationLower = destination.toLowerCase()
        // Folder name only, and substring-tolerant — Python matches
        // `loc.name.lower() == dest or dest in loc.name.lower()`, deliberately
        // *not* consulting display names. `getLocationByName` would be a wider
        // net; using it here would let the model travel to a location whose
        // folder does not exist, and the room key is built from the folder name.
        const match = dbLocations.find(
          (loc) =>
            loc.name.toLowerCase() === destinationLower ||
            loc.name.toLowerCase().includes(destinationLower),
        )

        if (!match) {
          return toolError(
            `Location '${destination}' does not exist. Use Task tool with location_designer ` +
              `to create it first.\n\nAvailable locations: ${
                dbLocations.length > 0 ? dbLocations.map((l) => l.name).join(', ') : 'none'
              }`,
          )
        }

        // A *fresh* room per visit, not the location's existing one. That is what
        // gives the arrival a clean conversation history: the room is the context
        // window, so revisiting a tavern must not replay the last visit's
        // dialogue at the model.
        const newRoom = createNewRoomForLocation(db, match)
        deps.rooms.setRoomMapping(
          worldName,
          RoomMappingService.locationToRoomKey(match.name),
          newRoom.id,
          [],
        )

        deps.status?.preConnectLocation?.(newRoom.id, match.id)

        setCurrentLocation(db, worldId, match.id)

        const fsState = deps.players.loadPlayerState(worldName)
        if (fsState) {
          fsState.currentLocation = match.name
          deps.players.savePlayerState(worldName, fsState)
        }

        const movedCharacters: string[] = []
        for (const charName of bringCharacters) {
          try {
            // Looked up world-wide, matching Python's `get_agent_by_name(db, name)`
            // with no world filter — a companion may have been created before the
            // world scoping existed.
            const agent = getAgentByName(db, charName)
            if (agent) {
              moveCharacterToLocation(db, agent.id, fromLocation?.id ?? null, match.id)
              movedCharacters.push(charName)
            } else {
              logger.warning(`Character '${charName}' not found, skipping`)
            }
          } catch (error) {
            logger.warning(`Failed to move character '${charName}': ${String(error)}`)
          }
        }

        try {
          createMessage(db, newRoom.id, {
            content: narration,
            role: 'assistant',
            agentId: requireAgentId(ctx),
          })
        } catch (error) {
          logger.error(`Failed to create travel narration: ${String(error)}`)
        }

        try {
          deps.rooms.saveSuggestions(worldName, [action1, action2])
        } catch (error) {
          logger.error(`Failed to save travel suggestions: ${String(error)}`)
        }

        try {
          deps.rooms.saveArrivalContext(worldName, {
            previousNarration: narration,
            triggeringAction: userAction,
            fromLocation: fromLocationName,
          })
        } catch (error) {
          logger.error(`Failed to save arrival context: ${String(error)}`)
        }

        let text = `**Traveled to:** ${match.name}\n- Position: (${match.positionX}, ${match.positionY})`
        if (movedCharacters.length > 0) {
          text += `\n- Companions: ${movedCharacters.join(', ')}`
        }
        text += '\n\n Narrative message created and displayed to player.'
        text += `\n Suggested actions: [${action1}] / [${action2}]`
        text += '\n Chat summary saved to world history.'
        text += '\n Arrival context saved for next action continuity.'

        return toolSuccess(text)
      } catch (error) {
        logger.error(`travel error: ${String(error)}`)
        return toolError(`Error during travel: ${String(error)}`)
      } finally {
        if (departureRoomId !== null) deps.status?.setSubAgentInactive?.(departureRoomId)
      }
    },
  )

  // `requireRoomId` is called for its check only: `travel` writes its narration
  // into the *new* room, but a context with no room at all has no departure
  // scene to summarise and should never have been offered the tool.
  requireRoomId(ctx)

  return [travel]
}

/**
 * `persist_location_design` — the Location Designer sub-agent's callback.
 *
 * Refuses to overwrite. A designer that re-runs and clobbers a location the
 * player has already visited would silently rewrite the map under them, so an
 * existing name is an error the sub-agent has to work around by picking another.
 */
export function createPersistLocationTool(
  ctx: ToolContext,
  deps: LocationToolDeps,
): SdkTool[] {
  const worldName = requireWorldName(ctx)
  const worldId = requireWorldId(ctx)

  const def = resolveTool(persistLocationDesignTool.name, ctx.groupName)
  if (!def || !deps.persistence) return []
  const persistenceFor = deps.persistence

  return [
    tool(
      persistLocationDesignTool.name,
      def.description,
      persistLocationDesignTool.inputSchema,
      async (args) => {
        try {
          const db = ctx.getDb()
          const dbLocations = getLocations(db, worldId)
          const nameLower = args.name.toLowerCase()

          const existing = dbLocations.find(
            (loc) =>
              loc.name.toLowerCase() === nameLower ||
              (loc.displayName !== null && loc.displayName.toLowerCase() === nameLower),
          )
          if (existing) {
            return toolError(`Location '${args.name}' already exists. Cannot overwrite.`)
          }

          const adjacentHints = args.adjacent_to ?? []

          const newLocationId = persistenceFor(db, worldId, worldName).createLocation({
            name: args.name,
            displayName: args.display_name,
            description: args.description,
            position: [args.position_x, args.position_y],
            adjacentHints,
            isStarting: args.is_starting,
          })

          // Adjacency is bidirectional and written from the pre-creation
          // snapshot, so a hint naming the location being created cannot link it
          // to itself.
          for (const adjacentName of adjacentHints) {
            const adjacent = dbLocations.find((loc) => loc.name === adjacentName)
            if (!adjacent) continue
            addAdjacentLocation(db, newLocationId, adjacent.id)
            addAdjacentLocation(db, adjacent.id, newLocationId)
          }

          // `is_starting` is informational during onboarding; the real starting
          // location is whatever `complete` is told, which is why this only logs.
          if (args.is_starting) {
            logger.info(`Location '${args.name}' marked as starting location candidate`)
          }

          return toolSuccess(
            `**Location Created:**\n` +
              `- Name: ${args.name}\n` +
              `- Position: (${args.position_x}, ${args.position_y})\n` +
              `- Is Starting: ${args.is_starting ? 'True' : 'False'}\n` +
              `- Description: ${args.description.slice(0, 200)}...`,
          )
        } catch (error) {
          logger.error(`persist_location_design error: ${String(error)}`)
          return toolError(`Error creating location: ${String(error)}`)
        }
      },
    ),
  ]
}
